[custom_ops] Support TensorList inputs/outputs

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124092
* #124089
* #124071
* #124066
* #124065
* #124064
* #123937
* #124063
* #124062
* __->__ #123615

We add a `supports_tensorlist` decorator that gives an autograd.Function
the ability to handle TensorLists.

Test Plan:
- custom_op_db tests

zou3519 committed:
[custom_ops] Support TensorList inputs/outputs

We add a `supports_tensorlist` decorator that gives an autograd.Function
the ability to handle TensorLists.

Test Plan:
- custom_op_db tests

[ghstack-poisoned]

vadimkantorov commented:
Related: https://github.com/pytorch/pytorch/issues/100249 :)

zou3519 committed:
Update on "[custom_ops] Support TensorList inputs/outputs"


We add a `supports_tensorlist` decorator that gives an autograd.Function
the ability to handle TensorLists.

Test Plan:
- custom_op_db tests

[ghstack-poisoned]

albanD commented on torch/_library/autograd.py:
```diff
@@ -31,8 +32,95 @@ def backward(ctx, *grads):
+
+import functools
+
+from ..utils import _pytree
```
Why is this at the bottom of the file?

zou3519 replied:
I hoped lintrunner would move it up but looks like it didn't

albanD commented on torch/_library/autograd.py:
```diff
@@ -31,8 +32,95 @@ def backward(ctx, *grads):
     return autograd_impl
+
+
+def supports_tensorlist(cls: Any) -> Any:
```
There is quite a bit of state here.
I think it would be good to have a few asserts to make sure that we only ever do apply -> fw -> bw -> apply -> fw -> bw -> etc
And that any double apply/bw/etc will raise a nice error saying this is not supported yet.

zou3519 replied:
That's a good catch. Let me think about if we can support the double apply situation -- that is probably going to be common.

zou3519 replied:
Updated with nice error messages. I also expand the number of things we support: we can support double apply. The secret is to put the input_spec/output_spec onto the ctx object so that each of the backward have their own input_spec/output_spec

albanD reviewed (commented):
Sounds fair. Small comment on more constraints to raise errors on weird behavior!

zou3519 committed:
Update on "[custom_ops] Support TensorList inputs/outputs"


We add a `supports_tensorlist` decorator that gives an autograd.Function
the ability to handle TensorLists.

Test Plan:
- custom_op_db tests

[ghstack-poisoned]

zou3519 committed:
Update on "[custom_ops] Support TensorList inputs/outputs"


We add a `supports_tensorlist` decorator that gives an autograd.Function
the ability to handle TensorLists.

Test Plan:
- custom_op_db tests

[ghstack-poisoned]

ezyang commented on torch/testing/_internal/custom_op_db.py:
```diff
@@ -177,97 +177,87 @@ def sample_inputs_numpy_view_copy(opinfo, device, dtype, requires_grad, **kwargs
     yield SampleInput(result, args=([2, 12],))
 
-@custom_ops.custom_op('_torch_testing::numpy_cat')
+@torch.library.custom_op('_torch_testing::numpy_cat', mutates_args=())
```
As best I can tell, the changes here are just... refactoring? You never use the new decorator here

zou3519 replied:
We register an autograd.Function to the Autograd dispatch key. In this PR, we apply the supports_tensorlist to said autograd.Function. This op, using the new custom ops API, would not have passed the new tests without the new decorator.

albanD commented on test/test_custom_ops.py:
```diff
@@ -2137,6 +2137,61 @@ def _(x, y):
+            @staticmethod
+            def forward(ctx, xs):
+                if len(xs) > 0:
+                    return Foo.apply(xs[1:])
```
Is doing a Foo.apply() during the "backward()" call ok? I would expect it to be a more likely thing users will do than this one.

zou3519 replied:
Yes, I can add a test for this

albanD commented on torch/_library/autograd.py:
```diff
@@ -31,8 +34,136 @@ def backward(ctx, *grads):
+    # NB: All calls to the autograd.Function.apply shares these variables
+    # We assume that only one call to .apply happens at a time. This means that
+    # you cannot call the autograd.Function recursively (e.g. from its own forward).
+    input_spec: Optional[spec_t] = None
+    output_spec: Optional[spec_t] = None
+    result_is_tuple = None
```
I think you can simplify most of these by:
- Returning specs/is_tuple as output to the forward. Any non-Tensor will just be ignored.
- Use a single nonlocal boolean flag to detect re-entrancy and keep the error checking

zou3519 replied:
I'm not sure this is cleaner because:
- we still need a nonlocal input_spec. Because we need to ferry this information from the .apply to the .forward
- adding additional outputs to forward means that we need to consider them in backward, which makes the backward function more complicated.

But I'm happy to try if you think this is better.

albanD replied:
I can buy that!

albanD commented on torch/_library/autograd.py:
```diff
@@ -31,8 +34,136 @@ def backward(ctx, *grads):
+    return cls
+
+
+def not_list_of_tensor(tree):
```
I don't recall the full details, but what's the link between our TensorList and tuple(Tensor,...) ?
I guess we convert tuple(Tensor, ...) to TensorList on the way in. But do we return actual lists?
I'm ok with being very strict here but that might be surprising to some users.

zou3519 replied:
For inputs, the pytorch dispatcher converts them to list
For outputs, the pytorch dispatcher converts tuples to list

I expect users to have trouble with the strictness (there are other problems, like not being able to pass None as the gradient for a TensorList), but I don't want to relax them in this PR (it's already getting complicated).

albanD reviewed (approved):
Ok sounds good!

zou3519 committed:
Update on "[custom_ops] Support TensorList inputs/outputs"


We add a `supports_tensorlist` decorator that gives an autograd.Function
the ability to handle TensorLists.

Test Plan:
- custom_op_db tests

[ghstack-poisoned]

zou3519 committed:
Update on "[custom_ops] Support TensorList inputs/outputs"


We add a `supports_tensorlist` decorator that gives an autograd.Function
the ability to handle TensorLists.

Test Plan:
- custom_op_db tests

[ghstack-poisoned]

zou3519 committed:
Update on "[custom_ops] Support TensorList inputs/outputs"


We add a `supports_tensorlist` decorator that gives an autograd.Function
the ability to handle TensorLists.

Test Plan:
- custom_op_db tests

[ghstack-poisoned]

zou3519 committed:
Update on "[custom_ops] Support TensorList inputs/outputs"


We add a `supports_tensorlist` decorator that gives an autograd.Function
the ability to handle TensorLists.

Test Plan:
- custom_op_db tests

[ghstack-poisoned]

zou3519 committed:
Update on "[custom_ops] Support TensorList inputs/outputs"


We add a `supports_tensorlist` decorator that gives an autograd.Function
the ability to handle TensorLists.

Test Plan:
- custom_op_db tests

[ghstack-poisoned]
